### ISSUE TITLE
start up is "java.lang.IndexOutOfBoundsException"

### ISSUE THREAD

**Issue by treeleafj** [spring-projects/spring-boot#12189]
application.yml:

```ymal
server:
  tomcat:
    basedir: /usr/boot/tomcat.${random.uuid}
```

```java
java.lang.IndexOutOfBoundsException
	at org.springframework.boot.loader.data.RandomAccessDataFile.getSubsection(RandomAccessDataFile.java:105)
	at org.springframework.boot.loader.jar.CentralDirectoryFileHeader.load(CentralDirectoryFileHeader.java:79)
	at org.springframework.boot.loader.jar.CentralDirectoryFileHeader.fromRandomAccessData(CentralDirectoryFileHeader.java:181)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:246)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:231)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:219)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:174)
	at org.springframework.boot.loader.jar.JarFile.getEntry(JarFile.java:209)
	at org.springframework.boot.loader.jar.JarFile.getJarEntry(JarFile.java:200)
	at org.springframework.boot.loader.jar.JarURLConnection.connect(JarURLConnection.java:103)
	at org.springframework.boot.loader.jar.JarURLConnection.getInputStream(JarURLConnection.java:170)
	at org.springframework.core.io.UrlResource.getInputStream(UrlResource.java:169)
	at com.baomidou.mybatisplus.spring.MybatisMapperRefresh.refresh(MybatisMapperRefresh.java:226)
	at com.baomidou.mybatisplus.spring.MybatisMapperRefresh.access$500(MybatisMapperRefresh.java:58)
	at com.baomidou.mybatisplus.spring.MybatisMapperRefresh$1.run(MybatisMapperRefresh.java:186)
	at java.lang.Thread.run(Thread.java:748)
java.lang.IndexOutOfBoundsException
	at org.springframework.boot.loader.data.RandomAccessDataFile.getSubsection(RandomAccessDataFile.java:105)
	at org.springframework.boot.loader.jar.CentralDirectoryFileHeader.load(CentralDirectoryFileHeader.java:79)
	at org.springframework.boot.loader.jar.CentralDirectoryFileHeader.fromRandomAccessData(CentralDirectoryFileHeader.java:181)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:246)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:231)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:219)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:174)
	at org.springframework.boot.loader.jar.JarFile.getEntry(JarFile.java:209)
	at org.springframework.boot.loader.jar.JarFile.getJarEntry(JarFile.java:200)
	at org.springframework.boot.loader.jar.JarURLConnection.connect(JarURLConnection.java:103)
	at org.springframework.boot.loader.jar.JarURLConnection.getInputStream(JarURLConnection.java:170)
	at org.springframework.core.io.UrlResource.getInputStream(UrlResource.java:169)
	at com.baomidou.mybatisplus.spring.MybatisMapperRefresh.refresh(MybatisMapperRefresh.java:226)
	at com.baomidou.mybatisplus.spring.MybatisMapperRefresh.access$500(MybatisMapperRefresh.java:58)
	at com.baomidou.mybatisplus.spring.MybatisMapperRefresh$1.run(MybatisMapperRefresh.java:186)
	at java.lang.Thread.run(Thread.java:748)
java.lang.IndexOutOfBoundsException
	at org.springframework.boot.loader.data.RandomAccessDataFile.getSubsection(RandomAccessDataFile.java:105)
	at org.springframework.boot.loader.jar.CentralDirectoryFileHeader.load(CentralDirectoryFileHeader.java:79)
	at org.springframework.boot.loader.jar.CentralDirectoryFileHeader.fromRandomAccessData(CentralDirectoryFileHeader.java:181)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:246)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:231)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:219)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:174)
	at org.springframework.boot.loader.jar.JarFile.getEntry(JarFile.java:209)
	at org.springframework.boot.loader.jar.JarFile.getJarEntry(JarFile.java:200)
	at org.springframework.boot.loader.jar.JarURLConnection.connect(JarURLConnection.java:103)
	at org.springframework.boot.loader.jar.JarURLConnection.getInputStream(JarURLConnection.java:170)
	at org.springframework.core.io.UrlResource.getInputStream(UrlResource.java:169)
	at com.baomidou.mybatisplus.spring.MybatisMapperRefresh.refresh(MybatisMapperRefresh.java:226)
	at com.baomidou.mybatisplus.spring.MybatisMapperRefresh.access$500(MybatisMapperRefresh.java:58)
	at com.baomidou.mybatisplus.spring.MybatisMapperRefresh$1.run(MybatisMapperRefresh.java:186)
	at java.lang.Thread.run(Thread.java:748)
java.lang.IndexOutOfBoundsException
	at org.springframework.boot.loader.data.RandomAccessDataFile.getSubsection(RandomAccessDataFile.java:105)
	at org.springframework.boot.loader.jar.CentralDirectoryFileHeader.load(CentralDirectoryFileHeader.java:79)
	at org.springframework.boot.loader.jar.CentralDirectoryFileHeader.fromRandomAccessData(CentralDirectoryFileHeader.java:181)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:246)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:231)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:219)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:174)
	at org.springframework.boot.loader.jar.JarFile.getEntry(JarFile.java:209)
	at org.springframework.boot.loader.jar.JarFile.getJarEntry(JarFile.java:200)
	at org.springframework.boot.loader.jar.JarURLConnection.connect(JarURLConnection.java:103)
	at org.springframework.boot.loader.jar.JarURLConnection.getInputStream(JarURLConnection.java:170)
	at org.springframework.core.io.UrlResource.getInputStream(UrlResource.java:169)
	at com.baomidou.mybatisplus.spring.MybatisMapperRefresh.refresh(MybatisMapperRefresh.java:226)
	at com.baomidou.mybatisplus.spring.MybatisMapperRefresh.access$500(MybatisMapperRefresh.java:58)
	at com.baomidou.mybatisplus.spring.MybatisMapperRefresh$1.run(MybatisMapperRefresh.java:186)
	at java.lang.Thread.run(Thread.java:748)

```

**Comment by wilkinsona** [comment]
Can you please provide a [sample](https://stackoverflow,com/help/mcve) that reproduces the problem?

**Comment by treeleafj** [comment]
use spring-boot- maven-plugin-1.5.9.release

application.yml:
```ymal
server:
  tomcat:
    basedir: /usr/boot/tomcat.${random.uuid}
```

core:
```java
@SpringBootApplication
public class Application {

    public static void main(String[] args) {
        SpringApplication.run(Application.class, args);
    }
}
```
package to demo.jar,

and  run "java -jar demo.jar" on CentOS is error:

console print:

```java
java.lang.IndexOutOfBoundsException
	at org.springframework.boot.loader.data.RandomAccessDataFile.getSubsection(RandomAccessDataFile.java:105)
	at org.springframework.boot.loader.jar.CentralDirectoryFileHeader.load(CentralDirectoryFileHeader.java:79)
	at org.springframework.boot.loader.jar.CentralDirectoryFileHeader.fromRandomAccessData(CentralDirectoryFileHeader.java:181)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:246)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:231)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:219)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:174)
	at org.springframework.boot.loader.jar.JarFile.getEntry(JarFile.java:209)
	at org.springframework.boot.loader.jar.JarFile.getJarEntry(JarFile.java:200)
	at org.springframework.boot.loader.jar.JarURLConnection.connect(JarURLConnection.java:103)
	at org.springframework.boot.loader.jar.JarURLConnection.getInputStream(JarURLConnection.java:170)
	at org.springframework.core.io.UrlResource.getInputStream(UrlResource.java:169)
	at com.baomidou.mybatisplus.spring.MybatisMapperRefresh.refresh(MybatisMapperRefresh.java:226)
	at com.baomidou.mybatisplus.spring.MybatisMapperRefresh.access$500(MybatisMapperRefresh.java:58)
	at com.baomidou.mybatisplus.spring.MybatisMapperRefresh$1.run(MybatisMapperRefresh.java:186)
	at java.lang.Thread.run(Thread.java:748)
java.lang.IndexOutOfBoundsException
	at org.springframework.boot.loader.data.RandomAccessDataFile.getSubsection(RandomAccessDataFile.java:105)
	at org.springframework.boot.loader.jar.CentralDirectoryFileHeader.load(CentralDirectoryFileHeader.java:79)
	at org.springframework.boot.loader.jar.CentralDirectoryFileHeader.fromRandomAccessData(CentralDirectoryFileHeader.java:181)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:246)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:231)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:219)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:174)
	at org.springframework.boot.loader.jar.JarFile.getEntry(JarFile.java:209)
	at org.springframework.boot.loader.jar.JarFile.getJarEntry(JarFile.java:200)
	at org.springframework.boot.loader.jar.JarURLConnection.connect(JarURLConnection.java:103)
	at org.springframework.boot.loader.jar.JarURLConnection.getInputStream(JarURLConnection.java:170)
	at org.springframework.core.io.UrlResource.getInputStream(UrlResource.java:169)
	at com.baomidou.mybatisplus.spring.MybatisMapperRefresh.refresh(MybatisMapperRefresh.java:226)
	at com.baomidou.mybatisplus.spring.MybatisMapperRefresh.access$500(MybatisMapperRefresh.java:58)
	at com.baomidou.mybatisplus.spring.MybatisMapperRefresh$1.run(MybatisMapperRefresh.java:186)
	at java.lang.Thread.run(Thread.java:748)
java.lang.IndexOutOfBoundsException
	at org.springframework.boot.loader.data.RandomAccessDataFile.getSubsection(RandomAccessDataFile.java:105)
	at org.springframework.boot.loader.jar.CentralDirectoryFileHeader.load(CentralDirectoryFileHeader.java:79)
	at org.springframework.boot.loader.jar.CentralDirectoryFileHeader.fromRandomAccessData(CentralDirectoryFileHeader.java:181)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:246)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:231)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:219)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:174)
	at org.springframework.boot.loader.jar.JarFile.getEntry(JarFile.java:209)
	at org.springframework.boot.loader.jar.JarFile.getJarEntry(JarFile.java:200)
	at org.springframework.boot.loader.jar.JarURLConnection.connect(JarURLConnection.java:103)
	at org.springframework.boot.loader.jar.JarURLConnection.getInputStream(JarURLConnection.java:170)
	at org.springframework.core.io.UrlResource.getInputStream(UrlResource.java:169)
	at com.baomidou.mybatisplus.spring.MybatisMapperRefresh.refresh(MybatisMapperRefresh.java:226)
	at com.baomidou.mybatisplus.spring.MybatisMapperRefresh.access$500(MybatisMapperRefresh.java:58)
	at com.baomidou.mybatisplus.spring.MybatisMapperRefresh$1.run(MybatisMapperRefresh.java:186)
	at java.lang.Thread.run(Thread.java:748)
java.lang.IndexOutOfBoundsException
	at org.springframework.boot.loader.data.RandomAccessDataFile.getSubsection(RandomAccessDataFile.java:105)
	at org.springframework.boot.loader.jar.CentralDirectoryFileHeader.load(CentralDirectoryFileHeader.java:79)
	at org.springframework.boot.loader.jar.CentralDirectoryFileHeader.fromRandomAccessData(CentralDirectoryFileHeader.java:181)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:246)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:231)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:219)
	at org.springframework.boot.loader.jar.JarFileEntries.getEntry(JarFileEntries.java:174)
	at org.springframework.boot.loader.jar.JarFile.getEntry(JarFile.java:209)
	at org.springframework.boot.loader.jar.JarFile.getJarEntry(JarFile.java:200)
	at org.springframework.boot.loader.jar.JarURLConnection.connect(JarURLConnection.java:103)
	at org.springframework.boot.loader.jar.JarURLConnection.getInputStream(JarURLConnection.java:170)
	at org.springframework.core.io.UrlResource.getInputStream(UrlResource.java:169)
	at com.baomidou.mybatisplus.spring.MybatisMapperRefresh.refresh(MybatisMapperRefresh.java:226)
	at com.baomidou.mybatisplus.spring.MybatisMapperRefresh.access$500(MybatisMapperRefresh.java:58)
	at com.baomidou.mybatisplus.spring.MybatisMapperRefresh$1.run(MybatisMapperRefresh.java:186)
	at java.lang.Thread.run(Thread.java:748)
````


i think, because "basedir: /usr/boot/tomcat.${random.uuid}"

${random.uuid}

**Comment by treeleafj** [comment]
This problem is random

**Comment by wilkinsona** [comment]
Thanks, but we need more than that. There’s code that you haven’t shared or even described in every stack trace shown above. If you’d like us to spend time investigating this please spend some time reading [this](https://stackoverflow.com/help/mcve) and then producing an appropriate sample.

**Comment by snicoll** [comment]
@treeleafj can you please try with `2.0.0.BUILD-SNAPSHOT`? I wonder if that's not a side effect on something we fixed yesterday (#12163).

**Comment by treeleafj** [comment]
Thanks,  However, the problem is random, there is no recurrence in the future, and if it happens again, I will tell you again
